Adds and implements ModelCustomizer

Grinderofl opened the pull request:
Provides a possible implementation for a way to alter modelbuilder from outside dbcontext.
Resolves: #2992 and #2805

dnfclas commented:
Hi __@Grinderofl__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

rowanmiller commented:
@AndriySvyryd can you take a quick look at this one and see if the design looks good. We may still hold off to post-7.0.0 to take this one.

AndriySvyryd commented:
@Grinderofl, sorry I didn't explain the purpose of the model cache key. It is to be used for the model cache in ModelSource. So `ConcurrentDictionary<Type, IModel> _models` would become `ConcurrentDictionary<ModelCacheKey, IModel> _models`, the default implementation of `ModelCacheKey.Equals` would just compare the stored `DbContext` types. Note that `ModelCacheKey` shouldn't store the `DbContext` instance to avoid leaking memory. This means that `IModelCustomizer.Customize` should take a `DbContext` instead of a `ModelCacheKey`

This would allow customizing the model differently depending on the `DbContext` instance data. E.g. one could put a property for a custom schema to be used for the particular `DbContext` instance, then implement custom `IModelCustomizer` that configures the model appropriately and a custom `ModelCacheKey` that would also use the schema for equals comparison.

We don't have a consensus oh what is the best way to add custom actions, so to make the PR simpler and more likely to go in sooner we can remove the requierement to add actions without modifying the service provider. This would make the default implementation `ModelCustomizer` very simple: it would just call `OnModelCreating`, so `ModelCustomizerOptions` and `ModelCustomizerOptionsSetup` can be removed. As long as `IModelCustomizer` is in DI, the users can choose and implement whatever makes more sense for their scenario. You could move `ModelCustomizerOptions` and `ModelCustomizerOptionsSetup` to the test to ensure that it can still be used like that.

Grinderofl commented:
That requirement of not modifying service provider did cause a lot of headaches. I've updated the PR and removed the options. I think it makes sense to just be able to swap out the service, as most other features can be altered this way.

AndriySvyryd commented on src/EntityFramework.Core/EntityFrameworkServiceCollectionExtensions.cs:
```diff
@@ -18,6 +18,7 @@
 using Microsoft.Data.Entity.ValueGeneration;
 using Microsoft.Extensions.Caching.Memory;
 using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options;
```
This shouldn't be needed

AndriySvyryd commented:
Could you also make the `ModelCacheKey` changes?

AndriySvyryd commented:
@Grinderofl We've discussed this as a team and decided that the interface should be split into these two:
```
    public interface IModelCacheKeyFactory
    {
        object Create([NotNull] DbContext context);
    }
```
```
    public interface IModelCustomizer
    {
        void OnModelCreating([NotNull] ModelBuilder modelBuilder, [NotNull] DbContext dbContext);
    }
```
`ModelCustomizer` can still implement both of them, but will need to be registered twice in DI

Could you make these changes, rebase on latest dev and squash the commits?

AndriySvyryd commented:
Thanks for your contribution!
